// @shopify/hydrogen/dist/esnext/foundation/HydrogenResponse/HydrogenResponse.server.js
import { CacheShort, generateCacheControlHeader, } from '../Cache/strategies/index.js';
import Redirect from '../Redirect/Redirect.client.js';
import React from 'react';
export class HydrogenResponse extends Response {
    wait = false;
    cacheOptions = CacheShort();
    status = 200;
    statusText = '';
    /**
     * Buffer the current response until all queries have resolved,
     * and prevent it from streaming back early.
     */
    doNotStream() {
        this.wait = true;
    }
    canStream() {
        return !this.wait;
    }
    cache(options) {
        if (options) {
            this.cacheOptions = options;
        }
        return this.cacheOptions;
    }
    get cacheControlHeader() {
        return generateCacheControlHeader(this.cacheOptions);
    }
    redirect(location, status = 307) {
        this.status = status;
        this.headers.set('location', location);
        // in the case of an RSC request, instead render a client component that will redirect
        return React.createElement(Redirect, { to: location });
    }
}
