// @shopify/cli-kit/dist/ui/executor.js
import { CustomInput } from './inquirer/input.js';
import { CustomAutocomplete } from './inquirer/autocomplete.js';
import { CustomSelect } from './inquirer/select.js';
import { CustomPassword } from './inquirer/password.js';
import inquirer from 'inquirer';
import fuzzy from 'fuzzy';
export async function run(question) {
    const questionName = question.name;
    return (await inquirer.prompt(question, { ...question.choices }))[questionName];
}
export function mapper(question) {
    switch (question.type) {
        case 'input':
            inquirer.registerPrompt('custom-input', CustomInput);
            return {
                ...question,
                type: 'custom-input',
            };
        case 'password':
            inquirer.registerPrompt('custom-password', CustomPassword);
            return {
                ...question,
                type: 'custom-password',
            };
        case 'select':
            inquirer.registerPrompt('custom-select', CustomSelect);
            return {
                ...question,
                type: 'custom-select',
                source: getAutompleteFilterType(),
            };
        case 'autocomplete':
            inquirer.registerPrompt('autocomplete', CustomAutocomplete);
            return {
                ...question,
                type: 'autocomplete',
                source: getAutompleteFilterType(),
            };
    }
}
function fuzzyFilter(answers, input = '') {
    return new Promise((resolve) => {
        resolve(fuzzy
            .filter(input, Object.values(answers), {
            extract(el) {
                return el.name;
            },
        })
            .map((el) => el.original));
    });
}
function containsFilter(answers, input = '') {
    return new Promise((resolve) => {
        resolve(Object.values(answers).filter((answer) => answer.name.includes(input)));
    });
}
function getAutompleteFilterType() {
    return process.env.SHOPIFY_USE_AUTOCOMPLETE_FILTER === 'fuzzy' ? fuzzyFilter : containsFilter;
}
